Introduce Lawnchair Advance Search

MrSluffy opened the pull request:
**Screenshot**
![BeFunky-collage (1)](https://github.com/LawnchairLauncher/lawnchair/assets/36076410/c6c50d15-5760-4969-bb53-53dbd172c106)



## Description
Initial Implementation advance search in Lawnchair
Enables search functionality across Files, Contacts, Settings Entry, History keyword and Suggestions on the Startpage.
<!-- Please include a summary of the change. Please also include relevant motivation and context. -->


Fixes #3740 <!-- optional -->

## Type of change
<!-- Replace :x: with :white_check_mark: to "check" the specified bullet -->

:x: General change (non-breaking change that doesn't fit the below categories like copyediting)
:white_check_mark: Bug fix (non-breaking change which fixes an issue)
:white_check_mark: New feature (non-breaking change which adds functionality)
:x: Breaking change (fix or feature that would cause existing functionality to not work as expected)

MrSluffy committed:
Introduce Lawnchair Device Search

Initial Implementation device search in Lawnchair:
Enables search functionality across Files, Contacts, and Suggestions on the Startpage.

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchItemBackground.kt:
```diff
@@ -18,10 +18,10 @@ class SearchItemBackground(
-    private val focusHighlight = ColorTokens.FocusHighlight.resolveColor(context)
-    private val groupHighlight = if (showBackground) ColorTokens.GroupHighlight.resolveColor(context) else 0
+    val focusHighlight = ColorTokens.FocusHighlight.resolveColor(context)
+    val groupHighlight = if (showBackground) ColorTokens.GroupHighlight.resolveColor(context) else 0
 
-    private val cornerRadii: FloatArray
+    val cornerRadii: FloatArray
```
Move these public fields under private ones.

MrSluffy replied:
this is intended as we used it in `SearchAdaptedItem` 
![image](https://github.com/LawnchairLauncher/lawnchair/assets/36076410/c6212af0-061e-49e6-aa33-bc046442d2f1)

Goooler replied:
Yeah, I mean't public fileds should be placed below private fields

MrSluffy replied:
Oh got it

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultIconRow.kt:
```diff
@@ -94,6 +102,17 @@ class SearchResultIconRow(context: Context, attrs: AttributeSet?) :
         setSubtitleText(target.searchAction?.subtitle, showDelimiter)
+        if (shouldHandleClick(target) && !isSmall) {
+            setOnClickListener {
+                target.searchAction?.intent?.let { it1 -> handleSearchTargetClick(context, it1) }
```
```suggestion
                target.searchAction?.intent?.let { intent -> handleSearchTargetClick(context, intent) }
```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultIconRow.kt:
```diff
@@ -94,6 +102,17 @@ class SearchResultIconRow(context: Context, attrs: AttributeSet?) :
+        if (isSuggestion) {
+            layoutParams.height = resources.getDimensionPixelSize(R.dimen.search_result_small_row_height)
+            setOnClickListener {
+                target.searchAction?.intent?.let { it1 -> handleSearchTargetClick(context, it1) }
```
```suggestion
                target.searchAction?.intent?.let { intent -> handleSearchTargetClick(context, intent) }
```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultText.kt:
```diff
@@ -0,0 +1,49 @@
+
+    override val titleText: CharSequence? get() = title.text
+
+    override fun launch(): Boolean {
+        return false
+    }
```
```suggestion
    override fun launch(): Boolean = false
```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultView.kt:
```diff
@@ -30,6 +38,16 @@ sealed interface SearchResultView {
+            targetCompat.layoutType != LayoutType.SMALL_ICON_HORIZONTAL_TEXT &&
+            targetCompat.resultType != SearchTargetCompat.RESULT_TYPE_SHORTCUT
+    }
+    fun handleSearchTargetClick(context: Context, searchTargetIntent: Intent) {
```
```suggestion

    fun handleSearchTargetClick(context: Context, searchTargetIntent: Intent) {
```

Goooler replied:
```suggestion

    fun handleSearchTargetClick(context: Context, searchTargetIntent: Intent) {
```

Goooler commented on lawnchair/src/app/lawnchair/search/SearchTargetCompat.kt:
```diff
@@ -262,7 +262,19 @@ data class SearchTargetCompat(
+        ],
+    )
     @Retention(AnnotationRetention.SOURCE)
     annotation class SearchResultType
```
We should convert this to Enum.

MrSluffy replied:
I think we can do it later converting to enum in separate commit changes since it's more on refactor, or maybe you can convert it and push in this branch?

Goooler replied:
Will do in a follow-up.

Goooler commented on lawnchair/src/app/lawnchair/search/SearchTargetUtils.kt:
```diff
@@ -7,14 +7,23 @@ import androidx.core.os.bundleOf
+private fun generateHashKey(input: String): String =
+    MessageDigest.getInstance("SHA-256")
+        .digest(input.toByteArray())
+        .joinToString("") { "%02x".format(it) }
```
```suggestion
        .joinToString() { "%02x".format(it) }
```

Goooler replied:
`""` is unnecessary?

MrSluffy replied:
Oh yeah, I forgot to remove that since I use that to joined key like "shortcut_"

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,193 @@
+    return withContext(Dispatchers.IO) {
+        try {
+            val url = URL("https://www.startpage.com/suggestions?q=$query&segment=startpage.lawnchair&partner=lawnchair&format=opensearch")
+            val connection = (url.openConnection() as HttpURLConnection).apply { requestMethod = "GET" }
```
Use Retrofit instead.

MrSluffy replied:
Oh yeah, just check we have retrofit will implement that

MrSluffy committed:
Apply reviewed suggestions by @Gooler

- Use retrofit for fetching suggestions

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -4,42 +4,56 @@ import android.content.Context
+import retrofit2.Response
+import retrofit2.Retrofit
+
+val json = Json { ignoreUnknownKeys = true }
```
Use `kotlinxJson` instead.

Goooler replied:
Use `kotlinxJson` instead.

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -4,42 +4,56 @@ import android.content.Context
+
+val retrofit: Retrofit = Retrofit.Builder()
+    .baseUrl("https://www.startpage.com/")
+    .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
```
And this https://github.com/LawnchairLauncher/lawnchair/blob/7aa0b7172c75000bfd3e113cc009f52dab9f979a/lawnchair/src/app/lawnchair/bugreport/KatbinService.kt#L20

MrSluffy committed:
use kotlinxJson

suphon-t commented:
Haven’t looked into the code yet, but I think suggestions should be under the search provider that it comes from instead. Having it under device search is a bit misleading because “device search” sounds like a local search.

pacjo commented:
Shouldn't suggestions be disabled by default, since it's sending data to 3rd party? Also ability to choose search provider would be nice.

Otherwise everything's working fine, albeit a bit slower then with only app search.

heckarman commented:
Can this be a replacement for Pixel Search app?

MrSluffy commented:
> Haven’t looked into the code yet, but I think suggestions should be under the search provider that it comes from instead. Having it under device search is a bit misleading because “device search” sounds like a local search.

Right, that make sense, let's add new category as Search provider / Suggestion provider in app drawer since it doesn't make sense to put it in dock since most of the dock func is redirect to browser / apps.

MrSluffy committed:
Separate suggestion

MrSluffy committed:
Merge branch '13-dev' into wide-search

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultRightLeftIcon.kt:
```diff
@@ -0,0 +1,142 @@
+    private fun launchApp(packageName: String, phoneNumber: String? = null) {
+        val packageManager = context.packageManager
+        val launchIntent = packageManager.getLaunchIntentForPackage(packageName)
+
+        if (launchIntent != null) {
+            if (packageName == defSmsAppInfo?.packageName) {
+                val smsIntent = Intent(Intent.ACTION_VIEW)
+                smsIntent.data = Uri.parse("smsto:$phoneNumber")
+                smsIntent.putExtra("address", phoneNumber)
+                handleSearchTargetClick(context, smsIntent)
+            } else if (packageName == defPhoneAppInfo?.packageName && phoneNumber != null) {
+                val phoneIntent = Intent(Intent.ACTION_DIAL)
+                phoneIntent.data = Uri.parse("tel:$phoneNumber")
+                handleSearchTargetClick(context, phoneIntent)
+            }
+        }
```
```suggestion

        launchIntent ?: return
            if (packageName == defSmsAppInfo?.packageName) {
                val smsIntent = Intent(Intent.ACTION_VIEW)
                smsIntent.data = Uri.parse("smsto:$phoneNumber")
                smsIntent.putExtra("address", phoneNumber)
                handleSearchTargetClick(context, smsIntent)
            } else if (packageName == defPhoneAppInfo?.packageName && phoneNumber != null) {
                val phoneIntent = Intent(Intent.ACTION_DIAL)
                phoneIntent.data = Uri.parse("tel:$phoneNumber")
                handleSearchTargetClick(context, phoneIntent)
            }
```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultRightLeftIcon.kt:
```diff
@@ -0,0 +1,142 @@
+            avatar.visibility = GONE
+            preview.visibility = VISIBLE
+        }
+
+        val layoutParams = LayoutParams(
+            LayoutParams.MATCH_PARENT,
+            if (isSmall) {
+                resources.getDimensionPixelSize(R.dimen.search_result_small_row_height)
+            } else {
+                resources.getDimensionPixelSize(R.dimen.search_result_files_row_height)
+            },
+        )
```
```suggestion
        val heightRes = if (isSmall) R.dimen.search_result_small_row_height else R.dimen.search_result_files_row_height
        val layoutParams = LayoutParams(
            LayoutParams.MATCH_PARENT,
            resources.getDimensionPixelSize(heightRes),
        )
```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultText.kt:
```diff
@@ -0,0 +1,47 @@
+
+    override val titleText: CharSequence? get() = title.text
+
+    override fun launch(): Boolean = false
```
```suggestion
    override fun launch(): Boolean = false

```

Goooler commented on lawnchair/src/app/lawnchair/allapps/SearchResultText.kt:
```diff
@@ -0,0 +1,47 @@
+        title.text = target.searchAction?.title
+        val layoutParams = LayoutParams(
+            LayoutParams.MATCH_PARENT,
+            when (title.text) {
+                SPACE -> resources.getDimensionPixelSize(R.dimen.space_layout_height)
+                SPACE_MINI -> resources.getDimensionPixelSize(R.dimen.space_layout_mini_height)
+                else -> resources.getDimensionPixelSize(R.dimen.search_result_text_height)
+            },
```
```kt
val res = when (title.text) {
  SPACE -> ...
  ...
}
resources.getDimensionPixelSize(res)
```

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+        val url = getStartPageUrl(suggestion)
+        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
+        val id = suggestion + url
+        val action = SearchActionCompat.Builder(
+            id,
+            suggestion,
+        )
```
```suggestion
        val action = SearchActionCompat.Builder(id, suggestion)
```

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+
+    internal fun getHeaderTarget(header: String): SearchTargetCompat {
+        val id = "header_$header"
+        val action = SearchActionCompat.Builder(
+            id,
+            header,
+        )
```
Same line like above.

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+    internal fun getMarketSearchItem(query: String): SearchTargetCompat? {
+        if (marketSearchComponent == null) return null
+        val id = "marketSearch:$query"
+        val action = SearchActionCompat.Builder(
+            id,
+            context.getString(R.string.all_apps_search_market_message),
+        )
```
Same.

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+        val url = getStartPageUrl(query)
+        val id = "browser:$query"
+        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
+        val action = SearchActionCompat.Builder(
+            id,
+            context.getString(R.string.all_apps_search_startpage_message),
+        )
```
Too.

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+            Intent.ACTION_VIEW,
+            contactUri,
+        )
+        val action = SearchActionCompat.Builder(
+            id,
+            info.name,
+        )
```
As well.

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+            fileIntent.setPackage(fileManagerPackageName)
+        }
+
+        val action = SearchActionCompat.Builder(
+            id,
+            info.name,
+        )
```
Also.

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,254 @@
+        return "com.google.android.apps.nbu.files"
+    }
+
+    private fun getPreviewIcon(info: FileInfo): Icon {
+        return when {
```
```suggestion
    private fun getPreviewIcon(info: FileInfo): Icon = when {
```

Goooler commented on lawnchair/src/app/lawnchair/search/LawnchairSearchAdapterProvider.kt:
```diff
@@ -51,23 +54,14 @@ class LawnchairSearchAdapterProvider(
-        return BaseAllAppsAdapter.ViewHolder(view)
-    }
-
+    ): BaseAllAppsAdapter.ViewHolder =
+        BaseAllAppsAdapter.ViewHolder(
+            layoutInflater.inflate(
+                layoutIdMap[viewType],
+                parent,
+                false,
+            ),
+        )
```
```suggestion
    ): BaseAllAppsAdapter.ViewHolder {
      val view = layoutInflater.inflate
      return BaseAllAppsAdapter.ViewHolder(view)
    }
```

Goooler commented on lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt:
```diff
@@ -55,20 +69,60 @@ sealed class LawnchairSearchAlgorithm(
+            SearchAdapterItem.createAdapterItem(target, background)
+        }
+    }
+    private fun findIndices(filtered: List<SearchTargetCompat>, layoutType: String): List<Int> {
```
```suggestion

    private fun findIndices(filtered: List<SearchTargetCompat>, layoutType: String): List<Int> {
```

Goooler commented on lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt:
```diff
@@ -55,20 +69,60 @@ sealed class LawnchairSearchAlgorithm(
+            else -> centerBackground
+        }
+    }
+    private fun getGroupedBackground(index: Int, indices: List<Int>): SearchItemBackground {
+        return when {
+            indices.size == 1 -> normalBackground
+            index == indices.first() -> topBackground
+            index == indices.last() -> bottomBackground
+            else -> centerBackground
+        }
```
```suggestion
    private fun getGroupedBackground(index: Int, indices: List<Int>): SearchItemBackground = when {
            indices.size == 1 -> normalBackground
            index == indices.first() -> topBackground
            index == indices.last() -> bottomBackground
            else -> centerBackground
```

Goooler commented on lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt:
```diff
@@ -92,6 +146,27 @@ sealed class LawnchairSearchAlgorithm(
     }
+
+    protected suspend fun performDeviceWideSearch(query: String, prefs: PreferenceManager): MutableList<SearchResult> = withContext(Dispatchers.IO) {
+        val results = ArrayList<SearchResult>()
```
Seems we can use `Sequence` here.

MrSluffy replied:
yes we can, when we implement max results option in separate commit since we define the result atm 
![image](https://github.com/LawnchairLauncher/lawnchair/assets/36076410/0802dec4-ef7a-49a9-a785-61a047506c0a)

Goooler commented on lawnchair/src/app/lawnchair/search/SearchTargetUtils.kt:
```diff
@@ -7,14 +7,23 @@ import androidx.core.os.bundleOf
+// the wrong action when click,
+// TODO remove when we properly manage creating search target
+private fun generateHashKey(input: String): String =
+    MessageDigest.getInstance("SHA-256")
+        .digest(input.toByteArray())
+        .joinToString { "%02x".format(it) }
```
you can do this easily with Okio.

MrSluffy replied:
is this fine? `ByteString.of(*MessageDigest.getInstance("SHA-256").digest(input.toByteArray`

Goooler replied:
Something like:
```kt
import okio.ByteString
import okio.ByteString.Companion.toByteString

fun ByteArray.sha1(separator: CharSequence = ""): String = toByteString().sha1().toHex(separator)

fun ByteArray.sha256(separator: CharSequence = ""): String =
  toByteString().sha256().toHex(separator)

fun ByteArray.md5(separator: CharSequence = ""): String = toByteString().md5().toHex(separator)

fun ByteArray.toHexString(separator: CharSequence = ""): String = joinToString(separator) {
  it.toInt().and(0xff).toString(16).padStart(2, '0')
}

private fun ByteString.toHex(separator: CharSequence) = toByteArray().toHexString(separator)
```

MrSluffy replied:
This is nice, but the current kinda fast, can you confirm how's this better the current one?

Goooler replied:
Little performance difference, but easy to maintain.

MrSluffy replied:
Right, I think we can do that when we have other use cases for now let's keep `ByteString.of(*MessageDigest.getInstance("SHA-256").digest(input.toByteArray`

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchDataInfo.kt:
```diff
@@ -0,0 +1,104 @@
+import app.lawnchair.search.data.MimeTypes.MIME_ZIP
+import com.android.launcher3.R
+
+object MimeTypes {
+    const val MIME_APP = "application/vnd.android.package-archive"
+    const val MIME_AUDIO = "audio/"
+    const val MIME_CSV = "text/comma-separated-values"
+    const val MIME_EXCEL = "application/vnd.ms-excel"
+    const val MIME_IMAGES = "image/"
+    const val MIME_PDF = "application/pdf"
+    const val MIME_PPT1 = "application/vnd.ms-powerpoint"
+    const val MIME_PPT2 = "application/vnd.openxmlformats-officedocument.presentationml.presentation"
+    const val MIME_SRT = "application/x-subrip"
+    const val MIME_TEXT = "text/"
+    const val MIME_VIDEO = "video/"
+    const val MIME_WORD = "application/msword"
+    const val MIME_ZIP = "application/zip"
+}
```
I can intro a more complete mime types mapping in a separate PR.

MrSluffy replied:
yeah that will be great

Goooler replied:
We can use the mapping from #3790 now.

MrSluffy replied:
done can you check

MrSluffy committed:
Apply changes code refine @Goooler

MrSluffy committed:
Fix styling

MrSluffy commented:
https://i.imgur.com/d9evRkx.png after spotlessApply https://i.imgur.com/IDvJeS1.png please revert it @Goooler

Goooler commented:
Ooops, mis-read them.

MrSluffy committed:
Merge branch '13-dev' into wide-search

MrSluffy committed:
Implement MIME Type Compat

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,206 @@
+
+                        val fileId = it.getLong(fileIdIndex)
+                        val displayName = it.getString(displayNameIndex)
+                        val mimeType = getMimeType(displayName) ?: it.getString(mimeTypeIndex)
```
You are using `displayName` here, it should be `extension`. And we have no need to query mime from MediaStore, just use `String.extension2MimeType()`. For extension, we can use:

```kt
internal val Path.extension: String?
  get() {
    val dotIndex = name.lastIndexOf(".")
    return if (dotIndex == -1) null else name.substring(dotIndex + 1)
  }

internal val Path.nameWithoutExtension: String get() = name.substringBeforeLast(".")
```

MrSluffy replied:
Oh right, but displayName have extension too

Goooler replied:
Explicit extension as a field, and generate mime from it.

MrSluffy replied:
Right are you able to do it? I have some stuff to do if not I'll do it later after work. Thanks

MrSluffy replied:
can you implement this? and push to this branch Thanks

Goooler replied:
Yeah, but I'm busy now, will take a look tonight.

MrSluffy replied:
bump, can you implement this so we can merge? as I'm planning working on merge A14 in this weekends Thanks

Goooler replied:
I'm starting this.

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -480,8 +485,16 @@
+    <string name="all_apps_search_result_files">From your files</string>
+
+    <string name="warn_contact_permission_content">To enable People search, allow lawnchair to read contact permission</string>
+    <string name="warn_files_permission_content">To enable Files search, allow lawnchair to read storage permission</string>
```
```suggestion
    <string name="warn_contact_permission_content">Requires contact permission</string>
    <string name="warn_files_permission_content">Requires storage permission</string>
```
I don't think we need this to be too wordy.

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -480,8 +485,16 @@
+    <string name="all_apps_search_startpage_message">Search on <xliff:g id="startpage">Startpage</xliff:g></string>
     <string name="error_no_market_or_browser_installed">No app store or browser installed.</string>
 
+    <string name="all_apps_search_result_suggestions">Suggestions</string>
```
```suggestion
    <string name="all_apps_search_result_suggestions">From the web</string>
```

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -337,6 +337,11 @@
+    <string name="perform_wide_search_title">Device Search</string>
+    <string name="perform_wide_search_file">Files</string>
+    <string name="pref_suggestion_title">Suggestions</string>
+    <string name="pref_advance_search_category">Advance Search</string>
```
```suggestion
    <string name="pref_advance_search_category">Advanced Search</string>
```

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -337,6 +337,11 @@
 
+    <string name="perform_wide_search_title">Device Search</string>
+    <string name="perform_wide_search_file">Files</string>
+    <string name="pref_suggestion_title">Suggestions</string>
```
```suggestion
    <string name="pref_suggestion_title">Web suggestions (via Startpage)</string>
```

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -337,6 +337,11 @@
+    <string name="perform_wide_search_file">Files</string>
+    <string name="pref_suggestion_title">Suggestions</string>
+    <string name="pref_advance_search_category">Advance Search</string>
+    <string name="perform_wide_search_description">Show your contacts and files all app search result, Once enabled option is available in App Drawer</string>
```
```suggestion
    <string name="perform_wide_search_description">Search your contacts and files in app drawer search. Enable in App Drawer settings.</string>
```

SuperDragonXD commented on lawnchair/src/app/lawnchair/ui/preferences/AppDrawerPreferences.kt:
```diff
@@ -115,6 +119,41 @@ fun AppDrawerPreferences() {
             }
         }
+        val isDeviceSearch = prefs.performWideSearchExperimental.get() && showDrawerSearchBar.state.value
+        ExpandAndShrink(visible = isDeviceSearch) {
```
```suggestion
        if (isDeviceSearch) {
```
Expanding and shrinking is unnecessary, because enabling the feature requires you to enable device search in a separate interface (experimental settings)

SuperDragonXD reviewed (commented):
Just a few string changes.

Regarding "suggestions", I've also clarified that the suggestions come from the web (startpage).

I'm gonna suggest more changes related to the UI in both preferences and the actual search interface later.

suphon-t commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,206 @@
+suspend fun findContactByName(context: Context, query: String, max: Int): List<ContactInfo> {
+    if (query.isEmpty() || query.isBlank()) return emptyList()
+    return withContext(Dispatchers.IO) {
+        suspendCoroutine { continuation ->
```
This doesn't need `suspendCoroutine`

suphon-t commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,206 @@
+suspend fun findByFileName(context: Context, query: String, max: Int): List<FileInfo> {
+    if (query.isEmpty() || query.isBlank()) return emptyList()
+    return withContext(Dispatchers.IO) {
+        suspendCoroutine { continuation ->
```
This too

suphon-t commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,206 @@
+suspend fun getStartPageSuggestions(query: String, max: Int): List<String> {
+    if (query.isEmpty() || query.isBlank() || max <= 0) return emptyList()
+
+    try {
```
Unnecessary outer try catch

MrSluffy replied:
Done, we move `if (query.isEmpty() || query.isBlank() || max <= 0) return emptyList()` inside try catch block

ps: no need to remove try catch as it will crash if user has no connection

suphon-t replied:
Oh right. Totally forgot that is a network call

suphon-t commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,206 @@
+
+val startPageService: StartPageService = retrofit.create(StartPageService::class.java)
+
+suspend fun getStartPageSuggestions(query: String, max: Int): List<String> {
```
Still think that this should be tied to the user's current search provider.
For example, a user that selected startpage as the provider should get suggestions from startpage, or a user that selected duckduckgo should get suggestions from duckduckgo.

MrSluffy replied:
I see, but we don't have endpoint for other provider, we only have from startpage. and also the current dock search implementation doesnt support your suggestions. it just redirect to browser when click.

suphon-t replied:
We could make tapping the dock open the drawer search instead. That way the user could type instantly without waiting for the browser to load.

About the other providers, what happens if the user clicks a startpage suggestion while using another provider?

MrSluffy replied:
Yes we can do that.

nothing, since it's not tied in the search provider in dock. I guess we can update it later on when we tied dock search bar to all appsearch

MrSluffy committed:
Apply suggestion changes by @suphon-t

MrSluffy committed:
Apply suggestion changes by @SuperDragonXD

Goooler committed:
Cleanup

Goooler committed:
Remodel FileInfo

Goooler committed:
Private retrofit

Goooler committed:
Clean up SearchRequest

Goooler committed:
Add FileExtensions

Goooler committed:
Optimize file search method

Goooler committed:
Merge branch '13-dev' into wide-search

Goooler committed:
Fix style

Goooler committed:
Rename

Goooler committed:
Optimize file opening

Goooler committed:
Add type checkers

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,198 @@
+        return ComponentKey(launchIntent.component, Process.myUserHandle())
+    }
+
+    private fun getPreviewIcon(info: FileInfo): Icon = when {
+        info.isMediaType -> BitmapFactory.decodeFile(info.path)?.let { Icon.createWithBitmap(it) }
+            ?: (
+                if (Utilities.ATLEAST_R) {
+                    MediaMetadataRetriever().run {
+                        setDataSource(info.path)
+                        val videoBitmap = frameAtTime
+                        release()
+                        videoBitmap?.let { Icon.createWithBitmap(it) }
+                    }
+                } else {
+                    null
+                }
+                ) ?: Icon.createWithResource(context, info.iconRes)
+        info.isUnknownType -> Icon.createWithBitmap(createTextBitmap(context, "U"))
+        else -> Icon.createWithResource(context, info.iconRes)
```
We can defer this bitmap fetching, just pass file path as the icon param, then load it with Coil.

Goooler commented:
Too much logic! So many optimizations have to be done.

Goooler committed:
Reformat AVDs

MrSluffy commented:
> Too much logic! So many optimizations have to be done.
https://katb.in/abahivihaqu

MrSluffy commented:
> Too much logic! So many optimizations have to be done.

ofc it will have much logic, point it out what to optimize so we can do it

Goooler commented:
Oh yes, I match both files and folders  when searching.

MrSluffy committed:
Merge branch '13-dev' into wide-search

MrSluffy committed:
Fixed crash java.lang.ClassCastException

error logs : https://katb.in/abahivihaqu

MrSluffy committed:
More fixes

MrSluffy committed:
Merge branch '13-dev' into wide-search

MrSluffy committed:
Surround try catch to give proper error logs

java.lang.NullPointerException
	at app.lawnchair.search.data.SearchRequestKt$findContactsByName$2.invokeSuspend(SearchRequest.kt:124)
	at app.lawnchair.search.data.SearchRequestKt$findContactsByName$2.invoke(Unknown Source:8)
	at app.lawnchair.search.data.SearchRequestKt$findContactsByName$2.invoke(Unknown Source:4)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:167)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
	at app.lawnchair.search.data.SearchRequestKt.findContactsByName(SearchRequest.kt:68)
	at app.lawnchair.search.LawnchairSearchAlgorithm$performDeviceWideSearch$2.invokeSuspend(LawnchairSearchAlgorithm.kt:152)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)

MrSluffy committed:
Fixed opened file instead of just redirect to file manager

MrSluffy committed:
Fixed crash opening folder

MrSluffy committed:
Initial implementation callback

MrSluffy committed:
Merge branch '13-dev' into wide-search

MrSluffy committed:
Add max results option for files, suggestion and people

MrSluffy committed:
Add initial settings entry

MrSluffy committed:
Merge branch '13-dev' into wide-search

MrSluffy committed:
Properly use folder mimetype

MrSluffy committed:
Fixed target icon styling

Goooler committed:
Cleanups

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,271 @@
+                        }
+                    } else {
+                        null
+                    } ?: Icon.createWithResource(context, fileInfo.iconRes)
```
Can we use `android.graphics.drawable.Icon#createWithFilePath` here? Or you can make icon param as `Any`, pass file path to it, then load with Coil.

MrSluffy replied:
Let me try it out

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,271 @@
+        val fileUri = when (info) {
+            is FileInfo -> Uri.withAppendedPath(
+                MediaStore.Files.getContentUri("external"),
+                info.fileId,
```
Why don't use file path to generate uri?

MrSluffy replied:
It doesn't open the file as it's intended, it just redirected/open the selected app

Goooler replied:
```kt
      val intent = Intent(Intent.ACTION_VIEW)
        .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        .addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
        .setDataAndType(path.path2Uri(), mime.mimeCompat)
      context?.startActivity(intent)
```

This should work, I just tried it on my file manager.

Goooler replied:
Probably cause I grant it with `ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION`?

MrSluffy replied:
Adding startActivity is useless here as it's already exists in SearchResultView

MrSluffy replied:
And it doesn't open the png/MP4 etc in respected app like Photos/gallery

Goooler replied:
Wired, it opens mp4 in my video player, and opens png in my photo viewer.

MrSluffy replied:
> Wired, it opens mp4 in my video player, and opens png in my photo viewer.

I'm pretty sure it doesn't, I tested in several apps and android 10 to 14 using emulator

Goooler replied:
Alright, will investigate it when I'm free of time.

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,293 @@
+                            jsonObject.put(CONTACT_ACCOUNT_NAME, accountName)
+                            jsonObject.put(CONTACT_ACCOUNT_TYPE, accountType)
+                            jsonObject.put(CONTACT_ACCOUNT_MIME, mimeType)
+                            jsonArray.put(jsonObject)
```
Use `buildJsonArray` and `buildJsonObject` instead.

Goooler commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -0,0 +1,293 @@
+
+            if (response.isSuccessful) {
+                val responseBody = response.body()?.string()
+                val suggestions = JSONArray(responseBody).optJSONArray(1)?.let { array ->
```
May KotlinX Json API instead.

MrSluffy replied:
it doesn't need to since the response is something like this
 `["popcorn",["popcorn", "popcorn instrumental", "popcorn sutton", "popcornflix" ]]` and the current is very straight forward and it's works as intended

MrSluffy commented on lawnchair/src/app/lawnchair/search/data/SearchRequest.kt:
```diff
@@ -36,53 +36,57 @@ private val retrofit = Retrofit.Builder()
-            return
-        }
+    val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        Log.e("Exception", "Error during suggestion retrieval: ${e.message}")
```
Have you tried this without data connection? I intended to surround it in try catch block as it's crashing when no data connection

Goooler replied:
May need to add `SupervisorJob()`.

MrSluffy replied:
> May need to add `SupervisorJob()`.
It's same

MrSluffy replied:
Please revert back surrounded try catch block thanks

Goooler committed:
Revert

MrSluffy committed:
Fixes

MrSluffy committed:
Apply suggestions by @Goooler

MrSluffy committed:
Fixed crash on external storage directory

MrSluffy committed:
Implement history keyword using built-in SearchRecentSuggestionsProvider

MrSluffy committed:
misc fixes

MrSluffy committed:
Add timeout 3 seconds

- As we don't how long gonna take to receive response in startpage

MrSluffy committed:
Misc fixes

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -332,11 +332,25 @@
+    <string name="perform_wide_search_file">Files</string>
+    <string name="pref_suggestion_title">Web suggestions (via <xliff:g id="startpage">Startpage</xliff:g>)</string>
+    <string name="pref_advance_search_category">Advanced Search</string>
+    <string name="perform_wide_search_description">Search your contacts and files in app drawer search. Enable in App Drawer settings.</string>
```
Actually, removing the last line should be fine.
```suggestion
    <string name="perform_wide_search_description">Search your contacts, files, and settings in app drawer search.</string>
```

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -481,9 +495,23 @@
+    <string name="warn_contact_permission_content">Requires contact permission</string>
+    <string name="warn_files_permission_content">Requires storage permission</string>
+
+    <string name="all_apps_search_result_files_description">Show files in search result</string>
```
```suggestion
    <string name="all_apps_search_result_files_description">Media, documents, folders, and more.</string>
```

SuperDragonXD commented on lawnchair/res/values/strings.xml:
```diff
@@ -481,9 +495,23 @@
+    <string name="warn_files_permission_content">Requires storage permission</string>
+
+    <string name="all_apps_search_result_files_description">Show files in search result</string>
+    <string name="all_apps_search_result_contacts_description">Show contacts in search result</string>
```
```suggestion
    <string name="all_apps_search_result_contacts_description">Contacts and more</string>
```

SuperDragonXD reviewed (commented):
I have some minor string suggestions.

Also, the current Preferences UI seems to be very convoluted and not-user friendly at all. I'll make a PR to fix this once this PR is merged.

MrSluffy committed:
Misc fixes

MrSluffy committed:
Merge branch '13-dev' into wide-search

Goooler commented on lawnchair/src/app/lawnchair/search/GenerateSearchTarget.kt:
```diff
@@ -0,0 +1,277 @@
+        return createSearchTarget(id, action, MARKET_STORE, extras)
+    }
+
+    internal fun getStartPageSearchItem(query: String): SearchTargetCompat {
+        val url = getStartPageUrl(query)
+        val id = "browser:$query"
+        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
+        val action = SearchActionCompat.Builder(id, context.getString(R.string.all_apps_search_startpage_message))
+            .setIcon(Icon.createWithResource(context, R.drawable.ic_startpage))
+            .setIntent(browserIntent)
+            .build()
+        val extras = bundleOf(
+            SearchResultView.EXTRA_HIDE_SUBTITLE to true,
+        )
+        return createSearchTarget(id, action, START_PAGE, extras)
+    }
```
Search item is always showing in the results, even if after turning `Web suggestions` off, is this an intended behavior?

MrSluffy replied:
No it doesn't show in my end, and this is not related to showing or showing of web suggestion, this only for wrapping the search result item , please check : https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt#L213

MrSluffy replied:
https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt#L191

MrSluffy replied:
maybe you got confused, there's a web suggestion and recent suggestion : https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/search/LawnchairSearchAlgorithm.kt#L226

Goooler replied:
<img width="300" alt="image" src="https://github.com/LawnchairLauncher/lawnchair/assets/10363352/3946b0f2-1366-4f76-ae49-633823f250f7">

<img width="300" alt="image" src="https://github.com/LawnchairLauncher/lawnchair/assets/10363352/065c45fb-d770-4b25-8727-192ac43e5885">

MrSluffy replied:
No, that is not web suggestion

MrSluffy replied:
and we don't have option for that

Goooler replied:
> and we don't have option for that

Is this intended?

MrSluffy replied:
This one is web suggestion
![image](https://github.com/LawnchairLauncher/lawnchair/assets/36076410/4f88c03a-23b8-4237-ac61-3032703a29ef)

MrSluffy replied:
> > and we don't have option for that
> 
> Is this intended?

Yes but if you want to add option you can do it here : https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/search/LawnchairAppSearchAlgorithm.kt#L149
